### PR TITLE
Add failsafe protections to Mode Takeoff

### PR DESCRIPTION
**Pull request by Hwurzburg**
Based off #23630 so may need changes as that one is reviewed....the only changes in mode_takeoff.cpp from that PR is the IF statement at line 133....
but this basically makes TAKEOFF have the same long failsafe protections as AUTO mode ....now long failsafe has no action if below TKOFF_LVL_ALT and gets recalled by TAKEOFF mode after reaching it and then will have the same behavior as AUTO in long failsafe...thanks to @peterbarker for brainstorming with me on this

### CODE REVIEW

File: ArduPlane/events.cpp [hunk 1]
```diff
@@ -18,6 +18,14 @@ bool Plane::failsafe_in_landing_sequence() const
     return false;
 }
 
+bool Plane::in_long_failsafe() const
+{
+   if (long_failsafe_active) {
+       return true;
+   }
+   return false;
```
Comment (IamPete1) [inline ArduPlane/events.cpp:26]:
```suggestion
   return long_failsafe_active;
```

**Comment by IamPete1** [comment]
We have three failsafe actions: RTL, glide and parachute. Is there any point in continuing the climb only to fire a parachute or start gliding at the end of it. In both cases I think that is a worse behaviour. That leaves RTL, there is already some min climb functionality in RTL so this extra climbing is a bit redundant. 

I would prefer that we override the min climb distance in RTL to be at least the take-off alt (if under the RTL alt) and then switch into RTL when the failsafe triggers as we would normally.

File: ArduPlane/events.cpp (hunk 1 above)
Comment (Hwurzburg) [reply ArduPlane/events.cpp]: done

**Comment by IamPete1** [comment]
This PR https://github.com/ArduPilot/ardupilot/pull/23682 is what we could do with a little rework, it adds the possibility of a postponed failsafe action.

File: ArduPlane/mode_takeoff.cpp [hunk 2]
```diff
@@ -150,6 +151,10 @@ void ModeTakeoff::update()
         plane.calc_nav_roll();
         plane.calc_nav_pitch();
         plane.calc_throttle();
+        //check if in long failsafe, if it is recall long failsafe now to get fs action via events call
+        if (plane.in_long_failsafe()) {
+        plane.failsafe_long_on_event(FAILSAFE_LONG, ModeReason::MODE_TAKEOFF_FAILSAFE);
```
Comment (IamPete1) [inline ArduPlane/mode_takeoff.cpp:156]: wont this repeated calling result in spamming the msg?`gcs().send_text(MAV_SEVERITY_WARNING, "%s Failsafe On: %s", (reason == ModeReason:: GCS_FAILSAFE) ? "GCS" : "RC Long", control_mode->name());`

File: ArduPlane/Plane.h [hunk 3]
```diff
@@ -983,6 +986,7 @@ class Plane : public AP_Vehicle {
     void failsafe_long_on_event(enum failsafe_state fstype, ModeReason reason);
     void failsafe_short_off_event(ModeReason reason);
     void failsafe_long_off_event(ModeReason reason);
+    bool in_long_failsafe() const; //returns true if in long_failsafe
```
Comment (tridge) [inline ArduPlane/Plane.h:989]: remove function and use variable directly

File: ArduPlane/mode_takeoff.cpp (hunk 2 above)
Comment (tridge) [inline ArduPlane/mode_takeoff.cpp:156]: clear the flag here before calling failsafe_long_on_event

**Review by tridge — CHANGES_REQUESTED** [review]
also need to clear this flag in mode::enter() in mode.cpp

File: ArduPlane/Plane.h [hunk 4]
```diff
@@ -217,6 +217,9 @@ class Plane : public AP_Vehicle {
     // are we doing calibration? This is used to allow heartbeat to
     // external failsafe boards during baro and airspeed calibration
     bool in_calibration;
+    
+    // are we currently in long failsafe>
+    bool long_failsafe_active;
```
Comment (tridge) [inline ArduPlane/Plane.h:222]: long_failsafe_pending would be a better name

**Review by peterbarker — COMMENTED** [review]
This *really* needs regression tests.

File: ArduPlane/Plane.h (hunk 4 above)
Comment (Hwurzburg) [reply ArduPlane/Plane.h]: done

File: ArduPlane/Plane.h (hunk 3 above)
Comment (Hwurzburg) [reply ArduPlane/Plane.h]: done

**Comment by Hwurzburg** [comment]
SITL tested....took some doing to get a long TAKEOFF flight phase in order to verify that GCS mode change clears the pending failsafe...but verified

**Review by tridge — APPROVED** [review]

**Comment by Hwurzburg** [comment]
thanks

**Pull request merged**